pallets/foreign-investments/src/impls/redeem.rs

```diff
@@ -0,0 +1,851 @@
+					..collected_swap
+				};
+				if amount_redeeming.is_zero() {
+					Ok(Self::ActiveSwapIntoForeignCurrency { swap: new_swap })
+				} else {
+					Ok(Self::RedeemingAndActiveSwapIntoForeignCurrency {
+						redeem_amount: amount_redeeming,
+						swap: new_swap,
+					})
+				}
+			}
+			RedeemingAndSwapIntoForeignDone { done_swap, .. } => {
+				if amount_redeeming.is_zero() {
+					Ok(Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+						swap: collected_swap,
+						done_amount: done_swap.amount,
+					})
+				} else {
+					Ok(
+						Self::RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+							redeem_amount: amount_redeeming,
+							swap: collected_swap,
+							done_amount: done_swap.amount,
+						},
+					)
+				}
+			}
+			RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+				swap,
+				done_amount,
+				..
+			} => {
+				let new_swap = Swap {
+					amount: swap.amount.ensure_add(collected_swap.amount)?,
+					..collected_swap
+				};
+				if amount_redeeming.is_zero() {
+					Ok(Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+						swap: new_swap,
+						done_amount,
+					})
+				} else {
+					Ok(
+						Self::RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+							redeem_amount: amount_redeeming,
+							swap: new_swap,
+							done_amount,
+						},
+					)
+				}
+			}
+			_ => Err(DispatchError::Other(
+				"Invalid pre redeem state when transitioning foreign collect",
+			)),
+		}
+	}
+}
+
+// Actual impl of transition
+impl<Balance, Currency> RedeemState<Balance, Currency>
+where
+	Balance: Clone + Copy + EnsureAdd + EnsureSub + Ord + Debug,
+	Currency: Clone + Copy + PartialEq + Debug,
+{
+	/// Handle `increase` transitions depicted by `msg::increase` edges in the
+	/// redeem state diagram:
+	/// * If the current state includes a processed investment, i.e. is either
+	///   `RedeemState::Invested` or `RedeemState::InvestedAnd(..)`, decreases
+	///   the invest amount and increases the redeeming amount. Throws if the
+	///   investment amount is exceeded as this reflects the max redeeming
+	///   amount.
+	/// * Else throws for incorrect pre state.
+	fn handle_increase(&self, amount: Balance) -> Result<Self, DispatchError> {
+		match self {
+			Self::Invested { invest_amount } => {
+				if &amount == invest_amount {
+					Ok(Self::NotInvestedAnd {
+						inner: Redeeming {
+							redeem_amount: amount,
+						},
+					})
+				} else {
+					Ok(Self::InvestedAnd {
+						invest_amount: invest_amount.ensure_sub(amount)?,
+						inner: Redeeming {
+							redeem_amount: amount,
+						},
+					})
+				}
+			}
+			Self::InvestedAnd {
+				invest_amount,
+				inner,
+			} => {
+				if &amount == invest_amount {
+					Ok(Self::NotInvestedAnd {
+						inner: inner.set_redeem_amount(amount)?,
+					})
+				} else {
+					Ok(Self::InvestedAnd {
+						invest_amount: invest_amount.ensure_sub(amount)?,
+						inner: inner.set_redeem_amount(amount)?,
```

Comment:
Wondering here if we do not need to increase the `reedem_amount` instead of set with the new `amount`

Reply:
Once again, great catch. You are right! Need to check tomorrow how that was not caught by current integration tests.